flora/flora.ino

```diff
@@ -100,123 +189,190 @@ bool getSensorData(BLEAddress pAddress, bool getBattery) {
+
+  bool batterySuccess = true;
+  if (readBattery) {
+    batterySuccess = readFloraBatteryCharacteristic(floraService, baseTopic);
+  }
+
+  return dataSuccess && batterySuccess;
+}
+
+bool processFloraDevice(BLEAddress floraAddress, char* deviceMacAddress, bool getBattery, int tryCount) {
+  Serial.print("Processing Flora device at ");
+  Serial.print(floraAddress.toString().c_str());
+  Serial.print(" (try ");
+  Serial.print(tryCount);
+  Serial.println(")");
+
+  // connect to flora ble server
+  BLEClient* floraClient = getFloraClient(floraAddress);
+  if (floraClient == nullptr) {
+    return false;
+  }
+
+  // connect data service
+  BLERemoteService* floraService = getFloraService(floraClient);
+  if (floraService == nullptr) {
+    floraClient->disconnect();
+    return false;
+  }
+
+  // process devices data
+  bool success = processFloraService(floraService, deviceMacAddress, getBattery);
+
+  // disconnect from device
+  floraClient->disconnect();
+
+  return success;
+}
+
+void hibernate() {
   esp_sleep_enable_timer_wakeup(SLEEP_DURATION * 1000000ll);
   Serial.println("Going to sleep now.");
+  delay(100);
   esp_deep_sleep_start();
 }
 
+void delayedHibernate(void *parameter) {
+  delay(EMERGENCY_HIBERNATE); // delay for five minutes
+  Serial.println("Something got stuck, entering emergency hibernate...");
+  hibernate();
+}
 
 void setup() {
+  // all action is done when device is woken up
   Serial.begin(115200);
   delay(1000);
-  
-  xTaskCreate(      taskDeepSleep,          /* Task function. */
-                    "TaskDeepSleep",        /* String with name of task. */
-                    10000,                  /* Stack size in words. */
-                    NULL,                   /* Parameter passed as input of the task */
-                    1,                      /* Priority of the task. */
-                    NULL);                  /* Task handle. */
-  
-  
-  Serial.println("Starting Flora client...");
-  BLEDevice::init("");
-
-  Serial.println("Connecting WiFi...");
-  WiFi.begin(wifi_ssid, wifi_password);
 
-  while (WiFi.status() != WL_CONNECTED) {
-    delay(500);
-    Serial.print(".");
-  }
+  // increase boot count
+  bootCount++;
 
-  Serial.println("");
-  Serial.println("WiFi connected");
+  // create a hibernate task in case something gets stuck
+  xTaskCreate(delayedHibernate, "hibernate", 4096, NULL, 1, &hibernateTaskHandle);
 
-  client.setServer(mqtt_server, 1883);
-  while (!client.connected()) {
-    Serial.print("Attempting MQTT connection...");
-    // Attempt to connect
-    if (client.connect("floraClient")) {
-      Serial.println("connected");
-    } else {
-      Serial.print("failed, rc=");
-      Serial.print(client.state());
-      Serial.println(" try again in 5 seconds");
-      // Wait 5 seconds before retrying
-      delay(5000);
+  Serial.println("Initialize BLE client...");
+  BLEDevice::init("");
+  BLEDevice::setPower(ESP_PWR_LVL_P7);
```

Comment:
uh, nice, didn't know that this feature exists :-)
does it actually help for better connectivity to the flora sensors?

Reply:
Cannot really tell since it doesn't connect reliably anyways. :-)